Interface to Krylov.jl solvers

amontoison opened the pull request:
Replace #3812
@glwagner @michel2323

I added all the routines needed by Krylov, but we probably want them implemented as kernels.

What I don't know is how we can overload the operators such that when we want to perform an operator-vector product with a `KrylovField`, we unwrap the `Field` in the custom vector type.
I think Gregory has an idea how we can do that.

The second question is how can we create a new `KrylovField` from an existing one? 
Can we create a function `KrylovField{T,F}(undef, n)`?  
I think we can't because `Field`s are like 3D arrays, and Krylov.jl should be modified to rely on a function like `similar` when a workspace is created.

glwagner commented on src/Solvers/krylov.jl:
```diff
@@ -0,0 +1,174 @@
+    return y
+end
+
+function Krylov.kcopy!(n::Int, y::KrylovField{T}, x::KrylovField{T}) where T <: FloatOrComplex
+    mx, nx, kx = size(x.field)
+    _x = x.field
+    _y = y.field
+    for i = 1:mx
+        for j = 1:nx
+            for k = 1:kx
+                _y[i,j,k] = _x[i,j,k]
+            end
+        end
+    end
+    return y
+end
```
```suggestion
Krylov.kcopy!(n::Int, y::KrylovField, x::KrylovField) = copyto!(y, x)
```

I think similar 1-liners may be defineable for most of the functions

glwagner commented on src/Solvers/krylov.jl:
```diff
@@ -0,0 +1,174 @@
+    return y
+end
+
+function Krylov.kfill!(x::KrylovField{T}, val::T) where T <: FloatOrComplex
+    mx, nx, kx = size(x.field)
+    _x = x.field
+    for i = 1:mx
+        for j = 1:nx
+            for k = 1:kx
+                _x[i,j,k] = val
+            end
+        end
+    end
+    return x
+end
```
```suggestion
Krylov.kfill!(x::KrylovField, val) = fill!(x.field, val)
```

It's better not to diagonalize the types / overconstrain the types in my opinion. For example `fill!(a, 0)` should definitely work even if `eltype(a) = Float64` or something.

glwagner commented:
Most of the functions are already defined so I don't think we need to rewrite the kernels right? We just need a couple 1-liners I think.

Also let's try not to overconstrain the types since this may create issues esp wrt autodifferentiation

amontoison commented:
@glwagner 
Do you have an implementation of `dot`, `norm`, `scal!`, `axpy!`, `axpby!` and `ref!` with a different name in Oceananigans.jl?
I think we need a kernel for these routines.

glwagner commented on src/Solvers/krylov.jl:
```diff
@@ -0,0 +1,152 @@
+Base.length(kf::KrylovField) = length(kf.field)
+Base.getindex(kf::KrylovField, i::Int) = getindex(kf.field, i)
+
+function Krylov.kdot(n::Int, x::KrylovField, y::KrylovField)
+    mx, nx, kx = size(x.field)
+    _x = x.field
+    _y = y.field
+    T = eltype(kf.field)
+    res = zero(T)
+    for i = 1:mx
+        for j = 1:nx
+            for k = 1:kx
+                res += _x[i,j,k] * _y[i,j,k]
+            end
+        end
+    end
+    return res
+end
```
```suggestion
Krylov.kdot(n::Int, x::KrylovField, y::KrylovField) = dot(x.field, y.field)
```

glwagner commented on src/Solvers/krylov.jl:
```diff
@@ -0,0 +1,152 @@
+    return res
+end
+
+function Krylov.knorm(n::Int, x::KrylovField)
+    mx, nx, kx = size(x.field)
+    _x = x.field
+    T = eltype(kf.field)
+    res = zero(T)
+    for i = 1:mx
+        for j = 1:nx
+            for k = 1:kx
+                res += _x[i,j,k]^2
+            end
+        end
+    end
+    return sqrt(res)
+end
```
```suggestion
Krylov.knorm(n::Int, x::KrylovField) = norm(x.field)
```

glwagner commented on src/Solvers/krylov.jl:
```diff
@@ -0,0 +1,152 @@
+    return sqrt(res)
+end
+
+function Krylov.kscal!(n::Int, s, x::KrylovField)
+    mx, nx, kx = size(x.field)
+    _x = x.field
+    for i = 1:mx
+        for j = 1:nx
+            for k = 1:kx
+                _x[i,j,k] = s * _x[i,j,k]
+            end
+        end
+    end
+    return x
+end
```
```suggestion
function Krylov.kscal!(n::Int, s, x::KrylovField)
    xp = parent(x.field)
    xp .*= s
    return x
end
```

will this be used by the solver?

amontoison replied:
Yes, but not all Krylov solvers.

glwagner commented:
> @glwagner Do you have an implementation of `dot`, `norm`, `scal!`, `axpy!`, `axpby!` and `ref!` with a different name in Oceananigans.jl? I think we need a kernel for these routines.

We have `dot` and `norm`. I think the others can be simple broadcasting and we can work on kernels once we have things working.

glwagner commented on src/Solvers/krylov.jl:
```diff
@@ -0,0 +1,125 @@
+end
+
+function Krylov.kaxpy!(n::Int, s, x::KrylovField, y::KrylovField)
+    mx, nx, kx = size(x.field)
+    _x = x.field
+    _y = y.field
+    for i = 1:mx
+        for j = 1:nx
+            for k = 1:kx
+                _y[i,j,k] += s * _x[i,j,k]
+            end
+        end
+    end
```
```suggestion
xp = parent(x.field)
yp = parent(y.field)
yp .+= s * xp
```

glwagner commented on src/Solvers/krylov.jl:
```diff
@@ -0,0 +1,125 @@
+end
+
+function Krylov.kaxpby!(n::Int, s, x::KrylovField, t, y::KrylovField)
+    mx, nx, kx = size(x.field)
+    _x = x.field
+    _y = y.field
+    for i = 1:mx
+        for j = 1:nx
+            for k = 1:kx
+                _y[i,j,k] = s * _x[i,j,k] + t * _y[i,j,k]
+            end
+        end
+    end
```
```suggestion
xp = parent(x.field)
yp = parent(y.field)
yp .= s .* xp .+ t .* yp
```

glwagner commented on src/Solvers/krylov.jl:
```diff
@@ -0,0 +1,125 @@
+                x_ijk = _x[i,j,k]
+                y_ijk = _y[i,j,k]
+                _x[i,j,k] = c       * x_ijk + s * y_ijk
+                _x[i,j,k] = conj(s) * x_ijk - c * y_ijk
```
doesn't this overwrite the statement above?

amontoison replied:
Line `77` should be `_y[i,j,k]`, it's a typo.

glwagner commented:
@amontoison here is a relatively complex example for you:

```julia
using Printf
using Statistics
using Oceananigans
using Oceananigans.Grids: with_number_type
using Oceananigans.BoundaryConditions: FlatExtrapolationOpenBoundaryCondition
using Oceananigans.Solvers: ConjugateGradientPoissonSolver, fft_poisson_solver
using Oceananigans.Simulations: NaNChecker
using Oceananigans.Utils: prettytime

N = 16
h, w = 50, 20
H, L = 100, 100
x = y = (-L/2, L/2)
z = (-H, 0)

grid = RectilinearGrid(size=(N, N, N); x, y, z, halo=(2, 2, 2), topology=(Bounded, Periodic, Bounded))

mount(x, y=0) = h * exp(-(x^2 + y^2) / 2w^2)
bottom(x, y=0) = -H + mount(x, y)
grid = ImmersedBoundaryGrid(grid, GridFittedBottom(bottom))

prescribed_flow = OpenBoundaryCondition(0.01)
extrapolation_bc = FlatExtrapolationOpenBoundaryCondition()
u_bcs = FieldBoundaryConditions(west = prescribed_flow,
                                east = extrapolation_bc)
                                #east = prescribed_flow)

boundary_conditions = (; u=u_bcs)
reduced_precision_grid = with_number_type(Float32, grid.underlying_grid)
preconditioner = fft_poisson_solver(reduced_precision_grid)
pressure_solver = ConjugateGradientPoissonSolver(grid; preconditioner, maxiter=1000, regularization=1/N^3)

model = NonhydrostaticModel(; grid, boundary_conditions, pressure_solver)
simulation = Simulation(model; Δt=0.1, stop_iteration=1000)
conjure_time_step_wizard!(simulation, cfl=0.5)

u, v, w = model.velocities
δ = ∂x(u) + ∂y(v) + ∂z(w)

function progress(sim)
    model = sim.model
    u, v, w = model.velocities
    @printf("Iter: %d, time: %.1f, Δt: %.2e, max|δ|: %.2e",
            iteration(sim), time(sim), sim.Δt, maximum(abs, δ))

    r = model.pressure_solver.conjugate_gradient_solver.residual
    @printf(", solver iterations: %d, max|r|: %.2e\n",
            iteration(model.pressure_solver), maximum(abs, r))
end

add_callback!(simulation, progress)

simulation.output_writers[:fields] =
    JLD2OutputWriter(model, model.velocities; filename="3831.jld2", schedule=IterationInterval(10), overwrite_existing=true)

run!(simulation)

using GLMakie

ds = FieldDataset("3831.jld2")
fig = Figure(size=(1000, 500))

n = Observable(1)
times = ds["u"].times
title = @lift @sprintf("time = %s", prettytime(times[$n]))

Nx, Ny, Nz = size(grid)
j = round(Int, Ny/2)
k = round(Int, Nz/2)
u_surface = @lift view(ds["u"][$n], :, :, k)
u_slice = @lift view(ds["u"][$n], :, j, :)

ax1 = Axis(fig[1, 1]; title = "u (xy)", xlabel="x", ylabel="y")
hm1 = heatmap!(ax1, u_surface, colorrange=(-0.01, 0.01), colormap=:balance)
Colorbar(fig[1, 2], hm1, label="m/s")

ax2 = Axis(fig[1, 3]; title = "u (xz)", xlabel="x", ylabel="z")
hm2 = heatmap!(ax2, u_slice, colorrange=(-0.01, 0.01), colormap=:balance)
Colorbar(fig[1, 4], hm2, label="m/s")

fig[0, :] = Label(fig, title)

record(fig, "3831.mp4", 1:length(times), framerate=10) do i
    n[] = i
end
```

I'll post some simpler examples for you too.

glwagner commented:
Here is a simple case:

```julia
using Printf
using Statistics
using Oceananigans
using Oceananigans.Grids: with_number_type
using Oceananigans.Solvers: ConjugateGradientPoissonSolver, fft_poisson_solver

N = 16
H, L = 100, 100
x = y = (-L/2, L/2)
z = (-H, 0)

grid = RectilinearGrid(size=(N, N, N); x, y, z, halo=(3, 3, 3), topology=(Periodic, Periodic, Bounded))

reduced_precision_grid = with_number_type(Float32, grid.underlying_grid)
preconditioner = fft_poisson_solver(reduced_precision_grid)
pressure_solver = ConjugateGradientPoissonSolver(grid; preconditioner, maxiter=1000, regularization=1/N^3)

model = NonhydrostaticModel(; grid, pressure_solver)

ui(x, y, z) = randn()
set!(model, u=ui, v=ui, w=ui)

simulation = Simulation(model; Δt=0.1, stop_iteration=100)
conjure_time_step_wizard!(simulation, cfl=0.5)

u, v, w = model.velocities
δ = ∂x(u) + ∂y(v) + ∂z(w)

function progress(sim)
    model = sim.model
    u, v, w = model.velocities
    @printf("Iter: %d, time: %.1f, Δt: %.2e, max|δ|: %.2e",
            iteration(sim), time(sim), sim.Δt, maximum(abs, δ))

    r = model.pressure_solver.conjugate_gradient_solver.residual
    @printf(", solver iterations: %d, max|r|: %.2e\n",
            iteration(model.pressure_solver), maximum(abs, r))
end

add_callback!(simulation, progress)

simulation.output_writers[:fields] =
    JLD2OutputWriter(model, model.velocities; filename="test.jld2", schedule=IterationInterval(10), overwrite_existing=true)

run!(simulation)
```

I can also help with visualization the results or you can try to use a similar code as above.

glwagner commented:
@xkykai you may be interested in following / helping with this work; @amontoison is replacing our in-house PCG implementation with something from Krylov, and he may be able to help with the problems we are currently having with the Poisson solver for certain problems on ImmersedBoundaryGrid

xkykai commented:
Yes, I've been revisiting the PCG solver, and had encountered the solver blowing up in the middle of a simulation of flow over periodic hill. I've tried for a bit trying to make it into a MWE but the problem was not showing up in simpler setups. This is also in a non open boundary condition setup so it is potentially separate from the issues brought up by others recently. My preliminary suspicion is that it could be a variable timestep issue and the model blows up when it is taking an almost machine precision timestep.

xkykai commented:
Yes, I've been revisiting the PCG solver, and had encountered the solver blowing up in the middle of a simulation of flow over periodic hill. I've tried for a bit trying to make it into a MWE but the problem was not showing up in simpler setups. This is also in a non open boundary condition setup so it is potentially separate from the issues brought up by others recently. My preliminary suspicion is that it could be a variable timestep issue and the model blows up when it is taking an almost machine epsilon timestep.

amontoison commented:
I think `Krylovable.jl` is a nice place to prototype the new solver.

amontoison commented:
Ok, let's do that 👍

francispoulin commented:
Any chance we will be able to use this to find eigenvalues of a matrix?  

If yes I have an idea for Kelvin Helmholtz instability example.  

If not, no problem.

amontoison commented:
You can it within a Rayleigh quotient method, which requires only a few lines of code if we have the Krylov solver.
-> https://en.wikipedia.org/wiki/Rayleigh_quotient_iteration

michel2323 commented:
Is there a showstopper for this PR @glwagner ?

glwagner commented:
No, I think this is great! It'd be nice to have a few tests before merging.

I also think it would be nice to see some benchmarks that demonstrate the superiority of Krylov here; we want to do this work anyways, and so its well-motivated to report that. It might stir up some excitement. Moreover (but it doesn't have to be done immediately) we should try to use a Krylov solver with FFT-based preconditioner for a nonhydrostatic problem in a complex domain. There are a few issues around with some tricky test problems that I think have stymied the current PCG solver. @xkykai can probably advise on this.

A final consideration is whether Krylov integration would be better implemented in an extension. I want to raise this because it is clearly feasible, but also say that my opinion is probably that it should _not_ be an extension. The argument for not doing an extension is: if Krylov is superior, I think we can motivate eliminating the other iterative solvers completely. In that case then, we want krylov solvers in src.

amontoison commented:
We have these benchmarks : https://github.com/glwagner/Krylovable.jl/pull/1
It proves that Krylov.jl performs as good as the current implementation in Oceananigans.jl.

glwagner commented:
> glwagner/Krylovable.jl#1

ah perfect! is it possible to reprroduce that with Oceananigans code? If not its ok. Just having some tests is good enough to keep going.

glwagner commented on src/Solvers/krylov.jl:
```diff
@@ -0,0 +1,87 @@
+    result = zeros(a.grid, 1)
+    Base.mapreducedim!(x -> x * x, +, result, conditional_a)
+    return CUDA.@allowscalar sqrt(first(result))
+end
```
I can't remember if I did this, but is there a reason to move this here? I think it better belongs in `Fields` module since it's not really specific to solvers or Krylov, for example.

amontoison replied:
You implemented this function, but you overloaded `Statistics.norm`, whereas (to the best of my knowledge) `norm` should be overloaded from `LinearAlgebra`.

amontoison replied:
I will move it back to `field.jl`.

xkykai commented:
I think by and large PCG solver is working fine currently, unless there are some issues we need to look into?

Here's a script that illustrates a pathological example which will go unstable if we use the FFT solver:

```julia
using Printf
using Dates
using Oceananigans
using Oceananigans.Operators
using Oceananigans.Units
using Oceananigans.Solvers
using Oceananigans.BuoyancyFormulations: buoyancy
using Oceananigans.Models.NonhydrostaticModels: ConjugateGradientPoissonSolver, FFTBasedPoissonSolver
using Random

const arch = GPU()
const time_stepper = :RungeKutta3
const advection = WENO()

const Nx = 128
const Nz = 32
const Lx = 128meters
const Lz = 32meters

const diffusivity = 1.0e-5
const Pr = 1

const f₀ = 1e-4
const stop_time = 1days

const xₗ, xᵣ = (-Lx/2, Lx/2)
const zₗ, zᵣ = (-Lz, 0.0)

const viscosity = Pr * diffusivity
rng_e = Xoshiro(123)

immersed_grid = begin
    underlying_grid = RectilinearGrid(
        arch,
        size = (Nx, Nz),
        x = (xₗ, xᵣ),
        z = (zₗ, zᵣ),
        topology = (Periodic, Flat, Bounded),
        halo = (4, 4)
    )

    dx = Lx / Nx

    width_x = 1/sqrt(Lz) * (xᵣ - 3 * dx)
    basin(x) = -Lz + (x/width_x)^2

    ImmersedBoundaryGrid(
        underlying_grid,
        GridFittedBottom(basin)
    )
end

coriolis = FPlane(f=f₀)

closure = ScalarDiffusivity(ν = viscosity, κ = diffusivity)

pressure_solver = nothing

# uncomment when using PCG solver to set preconditioner
# preconditioner = FFTBasedPoissonSolver(underlying_grid)
# preconditioner = nothing

# uncomment when using PCG solver
# reltol = abstol = 1e-7
# pressure_solver = ConjugateGradientPoissonSolver(immersed_grid, maxiter=100; reltol, abstol, preconditioner)

immersed_boundary_condition = ValueBoundaryCondition(0.)

model = NonhydrostaticModel(;
    grid = immersed_grid,
    timestepper = time_stepper,
    advection = advection,
    tracers = (:T, :S),
    coriolis = coriolis,
    buoyancy = SeawaterBuoyancy(),
    closure = closure,
    pressure_solver = pressure_solver
)

initial_T(x, z) = 20 + rand() * 1e-5
initial_S(x, z) = 35 + rand() * 1e-5

set!(model, T = initial_T)
set!(model, S = initial_S)

u, v, w = model.velocities

d = Field(∂x(u) + ∂z(w))

b_op = buoyancy(model)
b = Field(b_op)
compute!(b)

simulation = Simulation(model; Δt = 1e-4, stop_time = stop_time, minimum_relative_step = 1e-10)

time_wizard = TimeStepWizard(
    cfl = 0.6,
    max_change = 1.05,
    min_Δt = 1e-4,
    max_Δt = 10,
)

simulation.callbacks[:wizard] = Callback(
    time_wizard, IterationInterval(1)
)

wall_time = Ref(time_ns())

function progress(sim)
    if pressure_solver isa ConjugateGradientPoissonSolver
        pressure_iters = iteration(pressure_solver)
    else
        pressure_iters = 0
    end

    msg = @sprintf("Iter: %d, time: %s, Δt: %.4f, Poisson iters: %d",
                    iteration(sim), prettytime(time(sim)), sim.Δt, pressure_iters)

    elapsed = 1e-9 * (time_ns() - wall_time[])

    compute!(d)

    msg *= @sprintf(", max u: %6.3e, max w: %6.3e, max T: %6.3e, max S: %6.3e, max d: %6.3e, max pressure: %6.3e, wall time: %s",
                    maximum(abs, sim.model.velocities.u),
                    maximum(abs, sim.model.velocities.w),
                    maximum(abs, sim.model.tracers.T),
                    maximum(abs, sim.model.tracers.S),
                    maximum(abs, d),
                    maximum(abs, sim.model.pressures.pNHS),
                    prettytime(elapsed))

    @info msg
    wall_time[] = time_ns()

    return nothing
end

simulation.callbacks[:progress] = Callback(
    progress,
    IterationInterval(100),
    )

run!(simulation)
```

This is a parabolic well with constant temperature and salinity, seeded with minute noise. If we use the FFT solver the solution goes unstable due to flow divergence at the boundaries. If we use the PCG solver (which one can uncomment a few lines to do it) the solution behaves as expected. This could be a good test case to test the new Krylov solver?

amontoison commented:
@glwagner @xkykai 
Do you have some availability for a meeting this Thursday afternoon?
Michel and I are planning to work on it all day.
We would like to have an example with Krylov.jl + Oceananigans.jl for the Gordon Bell competition.

xkykai commented:
Tomorrow afternoon I would be available between 230pm and 4pm ET!

francispoulin commented:
Thank you @amontoison for your reply, and sorry for my late reply.  This is great to know!

I would be keen to test Krylov.jl to compute the eigenvalues for the Kelvin-Helmholtz example.  If people found this interesting, and useful, we might be able to add a bit to the example showing how this works (and presumably how much better it is than the power method).

I am happy to discuss if people wanted to but if not, no problem.

amontoison commented:
@glwagner @xkykai 
I backported the modifications of `Krylovable.jl` and added some unit tests.
The next steps depend on you.

amontoison commented:
> Thank you @amontoison for your reply, and sorry for my late reply. This is great to know!
> 
> I would be keen to test Krylov.jl to compute the eigenvalues for the Kelvin-Helmholtz example. If people found this interesting, and useful, we might be able to add a bit to the example showing how this works (and presumably how much better it is than the power method).
> 
> I am happy to discuss if people wanted to but if not, no problem.

@francispoulin We can have a look together once this PR is merged in Oceananigans.jl.

xkykai commented:
I noticed something odd with `KrylovSolver` not actually using the `preconditioner` that is passed into it, or perhaps I have missed something by using the API incorrectly. Below I provide a MWE:

```julia
using Oceananigans
using Oceananigans.BoundaryConditions: fill_halo_regions!
using Oceananigans.Solvers: FFTBasedPoissonSolver, solve!
using Oceananigans.Solvers: ConjugateGradientPoissonSolver, DiagonallyDominantPreconditioner, compute_laplacian!
using Oceananigans.Solvers: KrylovSolver
using LinearAlgebra
using Statistics
using Random
Random.seed!(123)

arch = CPU()
Nx = Ny = Nz = 64
topology = (Periodic, Periodic, Periodic)

x = y = z = (0, 2π)

grid = RectilinearGrid(arch; x, y, z, topology, size=(Nx, Ny, Nz))

δ = 0.1 # Gaussian width
gaussian(ξ, η, ζ) = exp(-(ξ^2 + η^2 + ζ^2) / 2δ^2)
Ngaussians = 17
Ξs = [2π * rand(3) for _ = 1:Ngaussians]

function many_gaussians(ξ, η, ζ)
    val = zero(ξ)
    for Ξ₀ in Ξs
        ξ₀, η₀, ζ₀ = Ξ₀
        val += gaussian(ξ - ξ₀, η - η₀, ζ - ζ₀)
    end
    return val
end

# Note: we use in-place transforms, so the RHS has to be AbstractArray{Complex{T}}.
# So, we first fill up "b" and then copy it into "bc = fft_solver.storage",
# which has the correct type.
b = CenterField(grid)
set!(b, many_gaussians)
parent(b) .-= mean(interior(b))

xpcg = CenterField(grid)

reltol = abstol = 1e-7

# preconditioner = nothing
preconditioner = DiagonallyDominantPreconditioner()
# preconditioner = FFTBasedPoissonSolver(grid)

preconditioner_str = preconditioner === nothing ? "nothing" : preconditioner isa FFTBasedPoissonSolver ? "FFT" : "DiagonallyDominant"

pcg_solver = ConjugateGradientPoissonSolver(grid, maxiter=1000; reltol, abstol, preconditioner)

# run it once for JIT
solve!(xpcg, pcg_solver.conjugate_gradient_solver, b)
xpcg .= 0
t_pcg = @timed solve!(xpcg, pcg_solver.conjugate_gradient_solver, b)
@info "PCG iteration $(pcg_solver.conjugate_gradient_solver.iteration), time $(t_pcg.time), preconditioner $(preconditioner_str)"

krylov_solver = KrylovSolver(compute_laplacian!; template_field=b, reltol, abstol, preconditioner)
xpcg_krylov = CenterField(grid)

# run it once for JIT
solve!(xpcg_krylov, krylov_solver, b)
xpcg_krylov .= 0
t_krylov = @timed solve!(xpcg_krylov, krylov_solver, b)
@info "Krylov.jl iteration $(krylov_solver.workspace.stats.niter), time $(t_krylov.time), preconditioner $(preconditioner_str)"

@info "Solution difference: $(norm(xpcg .- xpcg_krylov))"
```

Using `preconditioner = DiagonallyDominantPreconditioner()`, the output is as such:
```julia
[ Info: PCG iteration 89, time 0.366805, preconditioner DiagonallyDominant
[ Info: Krylov.jl iteration 177, time 0.560858, preconditioner DiagonallyDominant
[ Info: Solution difference: 5.144282484163763e-9
```

Using `preconditioner = nothing`, you get
```julia
[ Info: PCG iteration 178, time 0.6642343, preconditioner nothing
[ Info: Krylov.jl iteration 177, time 0.5929969, preconditioner nothing
[ Info: Solution difference: 4.983447757370835e-9
```

Using `preconditioner = FFTBasedPoissonSolver(grid)`, you'd get
```julia
[ Info: PCG iteration 1, time 0.0079233, preconditioner FFT
[ Info: Krylov.jl iteration 177, time 0.6844734, preconditioner FFT
[ Info: Solution difference: 2.4937669267347155e-8
```

It seems like the `preconditioner` is not implemented in `KrylovSolver` correctly?

amontoison commented:
@xkykai I just fixed it :+1: 

Note that we should find a good API for the operator / preconditioner such they support more than two arguments.
I was thinking of something like this such we can update the additional arguments needed by the linear operators:
```julia
function solve!(x, solver::KrylovSolver, b, args...; kwargs...)
    update_operator!(solver.op, args...)
    update_preconditioner(solver.preconditioner, args...)
    Krylov.solve!(solver.workspace, solver.op, KrylovField(b); M=solver.preconditioner,
                  atol=solver.abstol, rtol=solver.reltol, itmax=solver.maxiter, timemax=solver.maxtime, kwargs...)
    copyto!(x, solver.workspace.x.field)
    return x
end
```

xkykai commented on src/Solvers/krylov_solver.jl:
```diff
@@ -0,0 +1,133 @@
+    kc = Krylov.KrylovConstructor(kf)
+    workspace = Krylov.eval(Krylov.KRYLOV_SOLVERS[krylov_solver])(kc)
+
+    return KrylovSolver(arch, grid, op, workspace, krylov_solver, I, T(abstol), T(reltol), maxiter, maxtime)
```
How would one try to use one of the preconditioners we have? I tried `preconditioner = DiagonallyDominantPreconditioner()` and when running `solve!(xpcg_krylov, krylov_solver, b)` it throws something like
```julia
ERROR: MethodError: objects of type DiagonallyDominantPreconditioner are not callable
The object of type `DiagonallyDominantPreconditioner` exists, but no method is defined for this combination of argument types when trying to treat it as a callable object.
Stacktrace:
 [1] mul!(y::Oceananigans.Solvers.KrylovField{…}, A::Oceananigans.Solvers.KrylovOperator{…}, x::Oceananigans.Solvers.KrylovField{…})
   @ Oceananigans.Solvers c:\Users\xinle\MIT\Krylov_solver\Oceananigans.jl\src\Solvers\krylov_solver.jl:84
 [2] mulorldiv!(y::Oceananigans.Solvers.KrylovField{…}, P::Oceananigans.Solvers.KrylovOperator{…}, x::Oceananigans.Solvers.KrylovField{…}, ldiv::Bool)
   @ Krylov C:\Users\xinle\.julia\packages\Krylov\wgV9p\src\krylov_utils.jl:307
 [3] cg!(solver::Krylov.CgSolver{…}, A::Oceananigans.Solvers.KrylovOperator{…}, b::Oceananigans.Solvers.KrylovField{…}; M::Oceananigans.Solvers.KrylovOperator{…}, ldiv::Bool, radius::Float64, linesearch::Bool, atol::Float64, rtol::Float64, itmax::Int64, timemax::Float64, verbose::Int64, history::Bool, callback::Krylov.var"#837#845", iostream::Core.CoreSTDOUT)
   @ Krylov C:\Users\xinle\.julia\packages\Krylov\wgV9p\src\cg.jl:143
 [4] solve!(solver::Krylov.CgSolver{…}, A::Oceananigans.Solvers.KrylovOperator{…}, b::Oceananigans.Solvers.KrylovField{…}; M::Oceananigans.Solvers.KrylovOperator{…}, ldiv::Bool, radius::Float64, linesearch::Bool, atol::Float64, rtol::Float64, itmax::Int64, timemax::Float64, verbose::Int64, history::Bool, callback::Krylov.var"#837#845", iostream::Core.CoreSTDOUT)
   @ Krylov C:\Users\xinle\.julia\packages\Krylov\wgV9p\src\krylov_solve.jl:159
 [5] solve!(::Field{…}, ::KrylovSolver{…}, ::Field{…}; kwargs::@Kwargs{})
   @ Oceananigans.Solvers c:\Users\xinle\MIT\Krylov_solver\Oceananigans.jl\src\Solvers\krylov_solver.jl:129
 [6] solve!(::Field{…}, ::KrylovSolver{…}, ::Field{…})
   @ Oceananigans.Solvers c:\Users\xinle\MIT\Krylov_solver\Oceananigans.jl\src\Solvers\krylov_solver.jl:128
 [7] top-level scope
   @ c:\Users\xinle\MIT\Krylov_solver\Oceananigans.jl\krylov_preconditioner_mwe.jl:62
Some type information was truncated. Use `show(err)` to see complete types.
```

xkykai commented on src/Solvers/krylov_solver.jl:
```diff
@@ -0,0 +1,133 @@
+    # Linear operators
+    m = n = length(template_field)
+    op = KrylovOperator(T, m, n, linear_operator)
+    preconditioner = preconditioner == nothing ? I : KrylovOperator(T, m, n, preconditioner)
```
```suggestion
    preconditioner = preconditioner === nothing ? I : KrylovOperator(T, m, n, preconditioner)
```

xkykai commented:
Currently preconditioners are implemented with an API which is `precondition!(x, ::SomePreconditioner, rhs, args...) like
https://github.com/CliMA/Oceananigans.jl/blob/4921061ef6b102eb85677309c3ccf8f8645b0fb5/src/Solvers/conjugate_gradient_poisson_solver.jl#L141

Are you suggesting that `update_preconditioner!(solver.preconditioner, args...)` would make some preconditioner that follows a `linear_operator`-like function call so we can wrap it into `KrylovOperator`?

amontoison commented:
@xkykai @xkykai @michel2323 
Your `DiagonallyDominantPreconditioner` is not SPD, so you should not use it in a symmetric Krylov solver.
A symmetric Krylov solver requires a SPD Krylov solver, otherwise everything is wrong in terms of linear algebra.
The preconditioner should definite an elliptic norm such that we always have `x'Px >= 0`.
In this VERY specific case, CG can eat it but don't do that :smiling_face_with_tear: 
I added an absolute value in `heuristic_residual` and it fixed the issue.

For the `FFTBasedPoissonSolver` used as a preconditioner, we probably can't use in a symmetric solver because I don't see how it could represent an SPD matrix. I am also wondering if we can suppose that it represents a constant linear operator each time that we apply it.
If we are no reason to suspect that it is constant, we only have one solution, it is `fgmres` (flexible GMRES), it can support anything as a right preconditioner.

I also found a way to provide additional arguments to the operator `A` and the preconditioner `P`.
@xkykai Can probably try `bicgstab` + `FFTBasedPoissonSolver` in a problem that requires more than one iteration to confirm that please?

```julia
using Oceananigans
using Oceananigans.BoundaryConditions: fill_halo_regions!
using Oceananigans.Solvers: FFTBasedPoissonSolver, solve!
using Oceananigans.Solvers: ConjugateGradientPoissonSolver, DiagonallyDominantPreconditioner, compute_laplacian!
using Oceananigans.Solvers: KrylovSolver
using LinearAlgebra
using Statistics
using Random
Random.seed!(123)

arch = CPU()
Nx = Ny = Nz = 64
topology = (Periodic, Periodic, Periodic)

x = y = z = (0, 2π)

grid = RectilinearGrid(arch; x, y, z, topology, size=(Nx, Ny, Nz))

δ = 0.1 # Gaussian width
gaussian(ξ, η, ζ) = exp(-(ξ^2 + η^2 + ζ^2) / 2δ^2)
Ngaussians = 17
Ξs = [2π * rand(3) for _ = 1:Ngaussians]

function many_gaussians(ξ, η, ζ)
    val = zero(ξ)
    for Ξ₀ in Ξs
        ξ₀, η₀, ζ₀ = Ξ₀
        val += gaussian(ξ - ξ₀, η - η₀, ζ - ζ₀)
    end
    return val
end

# Note: we use in-place transforms, so the RHS has to be AbstractArray{Complex{T}}.
# So, we first fill up "b" and then copy it into "bc = fft_solver.storage",
# which has the correct type.
b = CenterField(grid)
set!(b, many_gaussians)
parent(b) .-= mean(interior(b))

xpcg = CenterField(grid)

reltol = abstol = 1e-7

# preconditioner = nothing
preconditioner = DiagonallyDominantPreconditioner()
# preconditioner = FFTBasedPoissonSolver(grid)

method = :cg
# methods = :bicgstab  # needed for `FFTBasedPoissonSolver`

preconditioner_str = preconditioner === nothing ? "nothing" : preconditioner isa FFTBasedPoissonSolver ? "FFT" : "DiagonallyDominant"

pcg_solver = ConjugateGradientPoissonSolver(grid, maxiter=1000; reltol, abstol, preconditioner)

# run it once for JIT
solve!(xpcg, pcg_solver.conjugate_gradient_solver, b)
xpcg .= 0
t_pcg = @timed solve!(xpcg, pcg_solver.conjugate_gradient_solver, b)
@info "PCG iteration $(pcg_solver.conjugate_gradient_solver.iteration), time $(t_pcg.time), preconditioner $(preconditioner_str)"

krylov_solver = KrylovSolver(compute_laplacian!; template_field=b, reltol, abstol, preconditioner, method)
xpcg_krylov = CenterField(grid)

# run it once for JIT
solve!(xpcg_krylov, krylov_solver, b)
xpcg_krylov .= 0
t_krylov = @timed solve!(xpcg_krylov, krylov_solver, b)
@info "Krylov.jl iteration $(krylov_solver.workspace.stats.niter), time $(t_krylov.time), preconditioner $(preconditioner_str)"

@info "Solution difference: $(norm(xpcg .- xpcg_krylov))"
```

xkykai commented:
> Your `DiagonallyDominantPreconditioner` is not SPD, so you should not use it in a symmetric Krylov solver.
> A symmetric Krylov solver requires a SPD Krylov solver, otherwise everything is wrong in terms of linear algebra.
> The preconditioner should definite an elliptic norm such that we always have `x'Px >= 0`.
> In this VERY specific case, CG can eat it but don't do that 🥲

ah yes :sweat_smile: I suppose as a general question for you, do you know how well do non-SPD preconditioners work for CG in practice? I understand that the theoretical guarantees are lost, in your experience does it cause a lot of issues in many cases?

Also as an aside I tried solving your many gaussians test case with FFT preconditioner and `:cg` in the Krylov solve and it throws `ERROR: DomainError with -1.844236867614985:` from the line
https://github.com/JuliaSmoothOptimizers/Krylov.jl/blob/cabd85e9ef5d51b63672e06f897dece01246633a/src/cg.jl#L146

When I switched `:cg` out to `:gmres` or `:fgmres` it throws 
```julia
ERROR: CanonicalIndexError: setindex! not defined for Oceananigans.Solvers.KrylovField{Float64, Field{Center, Center, Center, Nothing, RectilinearGrid{Float64, Periodic, Periodic, Periodic, Oceananigans.Grids.StaticVerticalDiscretization{OffsetArrays.OffsetVector{Float64, StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}}, OffsetArrays.OffsetVector{Float64, StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}}, Float64, Float64}, Float64, Float64, OffsetArrays.OffsetVector{Float64, StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}}, OffsetArrays.OffsetVector{Float64, StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}}, CPU}, Tuple{Colon, Colon, Colon}, OffsetArrays.OffsetArray{Float64, 3, Array{Float64, 3}}, Float64, FieldBoundaryConditions{BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Nothing}}, Nothing, Oceananigans.Fields.FieldBoundaryBuffers{Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing}}}
```
related to this line: https://github.com/JuliaSmoothOptimizers/Krylov.jl/blob/cabd85e9ef5d51b63672e06f897dece01246633a/src/gmres.jl#L218

> I also found a way to provide additional arguments to the operator `A` and the preconditioner `P`.
> @xkykai Can probably try `bicgstab` + `FFTBasedPoissonSolver` in a problem that requires more than one iteration to see it performs well please?

I am not 100% sure what you mean here. Did you mean using two preconditioners? How can I write that in the current `KrylovSolver` API?

amontoison commented:
> > Your `DiagonallyDominantPreconditioner` is not SPD, so you should not use it in a symmetric Krylov solver.
> > A symmetric Krylov solver requires a SPD Krylov solver, otherwise everything is wrong in terms of linear algebra.
> > The preconditioner should definite an elliptic norm such that we always have `x'Px >= 0`.
> > In this VERY specific case, CG can eat it but don't do that 🥲
> 
> ah yes 😅 I suppose as a general question for you, do you know how well do non-SPD preconditioners work for CG in practice? I understand that the theoretical guarantees are lost, in your experience does it cause a lot of issues in many cases?

You can't use non-SPD or non-SND preconditioners in CG.
Nothing is valid in the linear algebra if you do that.

> Also as an aside I tried solving your many gaussians test case with FFT preconditioner and `:cg` in the Krylov solve and it throws `ERROR: DomainError with -1.844236867614985:` from the line https://github.com/JuliaSmoothOptimizers/Krylov.jl/blob/cabd85e9ef5d51b63672e06f897dece01246633a/src/cg.jl#L146

Yes, it is because the preconditioner is not SPD.
> When I switched `:cg` out to `:gmres` or `:fgmres` it throws
>
> ```julia
> ERROR: CanonicalIndexError: setindex! not defined for Oceananigans.Solvers.KrylovField{Float64, Field{Center, Center, Center, Nothing, RectilinearGrid{Float64, Periodic, Periodic, Periodic, Oceananigans.Grids.StaticVerticalDiscretization{OffsetArrays.OffsetVector{Float64, StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}}, OffsetArrays.OffsetVector{Float64, StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}}, Float64, Float64}, Float64, Float64, OffsetArrays.OffsetVector{Float64, StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}}, OffsetArrays.OffsetVector{Float64, StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}}, CPU}, Tuple{Colon, Colon, Colon}, OffsetArrays.OffsetArray{Float64, 3, Array{Float64, 3}}, Float64, FieldBoundaryConditions{BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Nothing}}, Nothing, Oceananigans.Fields.FieldBoundaryBuffers{Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing}}}
> ```

I asked for `:bicgstab` and not the other ones :smiley:
`gmres` and `fgmres` should work once I released Krylov `v0.10`, I worked on it last weekend for the support here.
I added a few missing routines:
https://github.com/JuliaSmoothOptimizers/Krylov.jl/pull/980

> related to this line: https://github.com/JuliaSmoothOptimizers/Krylov.jl/blob/cabd85e9ef5d51b63672e06f897dece01246633a/src/gmres.jl#L218
> 
> > I also found a way to provide additional arguments to the operator `A` and the preconditioner `P`.

I mean that I fixed the issue with `args...`

> > @xkykai Can probably try `bicgstab` + `FFTBasedPoissonSolver` in a problem that requires more than one iteration to see it performs well please?
> 
> I am not 100% sure what you mean here. Did you mean using two preconditioners? How can I write that in the current `KrylovSolver` API?

I just wanted that try a problem with `method=:bicgstab` in the `KrylovSolver` + preconditioner based on an `FFTSolver` on a relevant problem where you can't apply the `FFTSolver` directly.

glwagner commented on src/Solvers/conjugate_gradient_poisson_solver.jl:
```diff
@@ -127,13 +127,12 @@ const FFTBasedPreconditioner = Union{FFTBasedPoissonSolver, FourierTridiagonalPo
-    arch = architecture(grid)
-    launch!(arch, grid, :xyz, subtract_and_mask!, p, grid, mean_p)
-
+    solve!(p, preconditioner, preconditioner.storage, -1e-8)
```
```suggestion
    shift = - sqrt(eps(eltype(r))) # to make the operator strictly negative definite
    solve!(p, preconditioner, preconditioner.storage, shift)
```

glwagner commented on src/Solvers/conjugate_gradient_poisson_solver.jl:
```diff
@@ -127,13 +127,12 @@ const FFTBasedPreconditioner = Union{FFTBasedPoissonSolver, FourierTridiagonalPo
-    launch!(arch, grid, :xyz, subtract_and_mask!, p, grid, mean_p)
-
+    solve!(p, preconditioner, preconditioner.storage, -1e-8)
+    p .= .-p
```
```suggestion
    p .*= -1
```

glwagner commented on src/Solvers/conjugate_gradient_poisson_solver.jl:
```diff
@@ -127,13 +127,12 @@ const FFTBasedPreconditioner = Union{FFTBasedPoissonSolver, FourierTridiagonalPo
-
+    solve!(p, preconditioner, preconditioner.storage, -1e-8)
+    p .= .-p
+    # mean_p = mean(p)
+    # grid = p.grid
+    # arch = architecture(grid)
+    # launch!(arch, grid, :xyz, subtract_and_mask!, p, grid, mean_p)
```
```suggestion
```

glwagner commented on src/Solvers/conjugate_gradient_poisson_solver.jl:
```diff
@@ -175,16 +174,15 @@ end
+                                                         2 * Ay⁻(i, j, k, grid) / (Ac(i, j, k, grid) + Ac(i, j-1, k, grid)) * r[i, j-1, k] -
+                                                         2 * Ay⁺(i, j, k, grid) / (Ac(i, j, k, grid) + Ac(i, j+1, k, grid)) * r[i, j+1, k] -
+                                                         2 * Az⁻(i, j, k, grid) / (Ac(i, j, k, grid) + Ac(i, j, k-1, grid)) * r[i, j, k-1] -
+                                                         2 * Az⁺(i, j, k, grid) / (Ac(i, j, k, grid) + Ac(i, j, k+1, grid)) * r[i, j, k+1])
```
this might help:

https://agupubs.onlinelibrary.wiley.com/doi/pdfdirect/10.1029/96JC02775

glwagner commented:
Any final thoughts @xkykai @tomchor @jagoosw ? Your work could be affected (not sure)

glwagner commented:
@Yixiao-Zhang

tomchor commented on src/Fields/abstract_field.jl:
```diff
@@ -124,6 +124,7 @@ for f in (:+, :-)
     @eval Base.$f(ϕ::AbstractField, ψ::AbstractArray) = $f(interior(ϕ), ψ)
 end
 
+
```
```suggestion
```

tomchor commented on src/Fields/field.jl:
```diff
@@ -587,7 +589,13 @@ const ReducedAbstractField = Union{XReducedAbstractField,
 # TODO: needs test
-Statistics.dot(a::Field, b::Field) = mapreduce((x, y) -> x * y, +, interior(a), interior(b))
+LinearAlgebra.dot(a::AbstractField, b::AbstractField) = mapreduce((x, y) -> x * y, +, interior(a), interior(b))
+# LinearAlgebra.norm(a::AbstractField) = mapreduce(x -> x * x, +, interior(a)) |> sqrt
```
```suggestion
```

xkykai commented:
Forgive me if I am using this incorrectly but when I try to use methods other than `:cg` it doesn't work? MWE is like

```julia
using Oceananigans
using Oceananigans.BoundaryConditions: fill_halo_regions!
using Oceananigans.Solvers: FFTBasedPoissonSolver, solve!
using Oceananigans.Solvers: ConjugateGradientPoissonSolver, DiagonallyDominantPreconditioner, compute_laplacian!
using Oceananigans.Solvers: KrylovSolver
using LinearAlgebra
using Statistics
using Random
Random.seed!(123)

arch = CPU()
Nx = Ny = Nz = 64
topology = (Periodic, Periodic, Periodic)

x = y = z = (0, 2π)

grid = RectilinearGrid(arch; x, y, z, topology, size=(Nx, Ny, Nz))

δ = 0.1 # Gaussian width
gaussian(ξ, η, ζ) = exp(-(ξ^2 + η^2 + ζ^2) / 2δ^2)
Ngaussians = 17
Ξs = [2π * rand(3) for _ = 1:Ngaussians]

function many_gaussians(ξ, η, ζ)
    val = zero(ξ)
    for Ξ₀ in Ξs
        ξ₀, η₀, ζ₀ = Ξ₀
        val += gaussian(ξ - ξ₀, η - η₀, ζ - ζ₀)
    end
    return val
end

# Note: we use in-place transforms, so the RHS has to be AbstractArray{Complex{T}}.
# So, we first fill up "b" and then copy it into "bc = fft_solver.storage",
# which has the correct type.
b = CenterField(grid)
set!(b, many_gaussians)
parent(b) .-= mean(interior(b))

xpcg = CenterField(grid)

reltol = abstol = 1e-7

# preconditioner = nothing
# preconditioner = DiagonallyDominantPreconditioner()
preconditioner = FFTBasedPoissonSolver(grid)

preconditioner_str = preconditioner === nothing ? "nothing" : preconditioner isa FFTBasedPoissonSolver ? "FFT" : "DiagonallyDominant"

pcg_solver = ConjugateGradientPoissonSolver(grid, maxiter=1000; reltol, abstol, preconditioner)

# run it once for JIT
solve!(xpcg, pcg_solver.conjugate_gradient_solver, b)
xpcg .= 0
t_pcg = @timed solve!(xpcg, pcg_solver.conjugate_gradient_solver, b)
@info "PCG iteration $(pcg_solver.conjugate_gradient_solver.iteration), time $(t_pcg.time), preconditioner $(preconditioner_str)"

method = :fgmres
krylov_solver = KrylovSolver(compute_laplacian!; template_field=b, reltol, abstol, preconditioner, method)
xpcg_krylov = CenterField(grid)

# run it once for JIT
solve!(xpcg_krylov, krylov_solver, b)
xpcg_krylov .= 0
t_krylov = @timed solve!(xpcg_krylov, krylov_solver, b)
@info "Krylov.jl iteration $(krylov_solver.workspace.stats.niter), time $(t_krylov.time), preconditioner $(preconditioner_str)"

@info "Solution difference: $(norm(xpcg .- xpcg_krylov))"
```
eg. here I use `:fgmres` and get error of 
```julia
ERROR: CanonicalIndexError: setindex! not defined for Oceananigans.Solvers.KrylovField{Float64, Field{Center, Center, Center, Nothing, RectilinearGrid{Float64, Periodic, Periodic, Periodic, Oceananigans.Grids.StaticVerticalDiscretization{OffsetArrays.OffsetVector{Float64, StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}}, OffsetArrays.OffsetVector{Float64, StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}}, Float64, Float64}, Float64, Float64, OffsetArrays.OffsetVector{Float64, StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}}, OffsetArrays.OffsetVector{Float64, StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}}, CPU}, Tuple{Colon, Colon, Colon}, OffsetArrays.OffsetArray{Float64, 3, Array{Float64, 3}}, Float64, FieldBoundaryConditions{BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Nothing}}, Nothing, Oceananigans.Fields.FieldBoundaryBuffers{Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing}}}
Stacktrace:
  [1] error_if_canonical_setindex(::IndexCartesian, A::Oceananigans.Solvers.KrylovField{Float64, Field{…}}, ::Int64)
    @ Base .\abstractarray.jl:1423
  [2] setindex!
    @ .\abstractarray.jl:1412 [inlined]
  [3] _setindex!
    @ .\abstractarray.jl:1448 [inlined]
  [4] setindex!
    @ .\abstractarray.jl:1413 [inlined]
  [5] macro expansion
    @ .\broadcast.jl:973 [inlined]
  [6] macro expansion
    @ .\simdloop.jl:77 [inlined]
  [7] copyto!
    @ .\broadcast.jl:972 [inlined]
  [8] copyto!
    @ .\broadcast.jl:925 [inlined]
  [9] materialize!
    @ .\broadcast.jl:883 [inlined]
 [10] materialize!
    @ .\broadcast.jl:880 [inlined]
 [11] kdivcopy!(n::Int64, y::Oceananigans.Solvers.KrylovField{…}, x::Oceananigans.Solvers.KrylovField{…}, s::Float64)
    @ Krylov C:\Users\xinle\.julia\packages\Krylov\3JTOh\src\krylov_utils.jl:334
 [12] fgmres!(solver::Krylov.FgmresSolver{…}, A::Oceananigans.Solvers.KrylovOperator{…}, b::Oceananigans.Solvers.KrylovField{…}; M::Oceananigans.Solvers.KrylovPreconditioner{…}, N::UniformScaling{…}, ldiv::Bool, restart::Bool, reorthogonalization::Bool, atol::Float64, rtol::Float64, itmax::Int64, timemax::Float64, verbose::Int64, history::Bool, callback::Krylov.var"#1032#1040", iostream::Core.CoreSTDOUT)
    @ Krylov C:\Users\xinle\.julia\packages\Krylov\3JTOh\src\fgmres.jl:224
 [13] fgmres!
    @ C:\Users\xinle\.julia\packages\Krylov\3JTOh\src\fgmres.jl:115 [inlined]
 [14] solve!
    @ C:\Users\xinle\.julia\packages\Krylov\3JTOh\src\krylov_solve.jl:159 [inlined]
 [15] #solve!#54
    @ c:\Users\xinle\MIT\Krylov_solver\Oceananigans.jl\src\Solvers\krylov_solver.jl:145 [inlined]
 [16] solve!(::Field{…}, ::KrylovSolver{…}, ::Field{…})
    @ Oceananigans.Solvers c:\Users\xinle\MIT\Krylov_solver\Oceananigans.jl\src\Solvers\krylov_solver.jl:142
 [17] top-level scope
    @ c:\Users\xinle\MIT\Krylov_solver\Oceananigans.jl\krylov_preconditioner_mwe.jl:63
Some type information was truncated. Use `show(err)` to see complete types.
```
I've tried adding this Krylov branch https://github.com/JuliaSmoothOptimizers/Krylov.jl/commit/cabd85e9ef5d51b63672e06f897dece01246633a but it doesn't seem to solve the issue.

Ignore me if this is not the case, I might have gotten confused about the right version of packages to use due to the merge conflict on `Project.toml`.

amontoison commented:
@xkykai You will need the release `0.10` of Krylov.jl for some solvers.
I will release it and quite soon.
`:cg` and `:bicgstab` work for sure now.

Yixiao-Zhang commented:
It might be off-topic, but I want to show how I treat the non-SPD matrices.

I think that shifting the mean value of `r` \(instead of `p`\) to zero at each iteration enables CG work for non-SPD matrices. I include the following patch in all my simulations that use the immersed boundary condition, and so far they been running smoothly without any numerical issues.

Without this patch, due to round-off errors, the mean value of `r` will drift away from zero at each iteration, and then `Ax = r` will not have any solution.

```Julia
function update_solution_and_residuals!(x, r, q, p, α)
    xp = parent(x)
    rp = parent(r)
    qp = parent(q)
    pp = parent(p)

    xp .+= α .* pp
    rp .-= α .* qp

    mean_r = mean(r)
    grid = r.grid
    arch = architecture(grid)
    launch!(arch, grid, :xyz, subtract_and_mask!, r, grid, mean_r)

    return nothing
end

```

amontoison commented:
@xkykai Do you just solve (A + a I) x = b instead where `a = mean_r` ?
I can add an option an Krylov.jl to directly provide this shift. It is quite common is symmetric Krylov solvers.

tomchor commented on src/Solvers/krylov_solver.jl:
```diff
@@ -0,0 +1,173 @@
+Krylov.kcopy!(n::Integer, y::KrylovField{T}, x::KrylovField{T}) where T <: FloatOrComplex = copyto!(y.field, x.field)
+Krylov.kfill!(x::KrylovField{T}, val::T) where T <: FloatOrComplex = fill!(x.field, val)
+
+# Only needed by the Krylov solver MINRES-QLP.
+# We can implement a kernel if we need it.
+#
+# function Krylov.kref!(n::Integer, x::KrylovField{T}, y::KrylovField{T}, c::T, s::T) where T <: FloatOrComplex
+#     mx, nx, kx = size(x.field)
+#     _x = x.field
+#     _y = y.field
+#     for i = 1:mx
+#         for j = 1:nx
+#             for k = 1:kx
+#                 x_ijk = _x[i,j,k]
+#                 y_ijk = _y[i,j,k]
+#                 _x[i,j,k] = c       * x_ijk + s * y_ijk
+#                 _y[i,j,k] = conj(s) * x_ijk - c * y_ijk
+#             end
+#         end
+#     end
+#     return x, y
+# end
```
Do we wanna keep this comment?

glwagner replied:
i'd say no, @amontoison want to delete the comments? They'll be in history if we need them

tomchor replied:
```suggestion
```

amontoison replied:
I removed it. I will add a dedicated kernel in the future if we need MINRES-QLP.

tomchor commented on test/test_krylov_solver.jl:
```diff
@@ -0,0 +1,71 @@
+    extrema_tolerance = 1e-12
+    std_tolerance = 1e-13
+
+    CUDA.@allowscalar begin
+        @test minimum(abs, interior(∇²ϕ_solution) .- interior(∇²ϕ)) < extrema_tolerance
+        @test maximum(abs, interior(∇²ϕ_solution) .- interior(∇²ϕ)) < extrema_tolerance
+        @test          std(interior(∇²ϕ_solution) .- interior(∇²ϕ)) < std_tolerance
+
+        @test   minimum(abs, interior(ϕ_solution) .- interior(ϕ_truth)) < extrema_tolerance
+        @test   maximum(abs, interior(ϕ_solution) .- interior(ϕ_truth)) < extrema_tolerance
+        @test            std(interior(ϕ_solution) .- interior(ϕ_truth)) < std_tolerance
+    end
```
I thought we were trying to avoid scalar operations. This can't be written without it?

glwagner replied:
looks like here we simply need to remove `interior` and `.`

glwagner replied:
well but also there are not scalar ops

tomchor commented:
I'm a bit lost on the overall purpose of this PR. Are Krylov solvers going to be used for iterative methods when solving for pressure?

glwagner commented on test/test_krylov_solver.jl:
```diff
@@ -0,0 +1,71 @@
+    extrema_tolerance = 1e-12
+    std_tolerance = 1e-13
+
+    CUDA.@allowscalar begin
```
```suggestion
```

glwagner commented on test/test_krylov_solver.jl:
```diff
@@ -0,0 +1,71 @@
+    extrema_tolerance = 1e-12
+    std_tolerance = 1e-13
+
+    CUDA.@allowscalar begin
+        @test minimum(abs, interior(∇²ϕ_solution) .- interior(∇²ϕ)) < extrema_tolerance
+        @test maximum(abs, interior(∇²ϕ_solution) .- interior(∇²ϕ)) < extrema_tolerance
+        @test          std(interior(∇²ϕ_solution) .- interior(∇²ϕ)) < std_tolerance
+
+        @test   minimum(abs, interior(ϕ_solution) .- interior(ϕ_truth)) < extrema_tolerance
+        @test   maximum(abs, interior(ϕ_solution) .- interior(ϕ_truth)) < extrema_tolerance
+        @test            std(interior(ϕ_solution) .- interior(ϕ_truth)) < std_tolerance
+    end
```
```suggestion
    @test minimum(abs, interior(∇²ϕ_solution) .- interior(∇²ϕ)) < extrema_tolerance
    @test maximum(abs, interior(∇²ϕ_solution) .- interior(∇²ϕ)) < extrema_tolerance
    @test          std(interior(∇²ϕ_solution) .- interior(∇²ϕ)) < std_tolerance

    @test   minimum(abs, interior(ϕ_solution) .- interior(ϕ_truth)) < extrema_tolerance
    @test   maximum(abs, interior(ϕ_solution) .- interior(ϕ_truth)) < extrema_tolerance
    @test            std(interior(ϕ_solution) .- interior(ϕ_truth)) < std_tolerance
```

scalar ops here would be a bug, so we def don't want to write `allowscalar`

amontoison replied:
I copy-pasted from an another test with PCG (the one that is implemented in Oceananigans.jl).

amontoison commented:
I can't merge this PR because of the failing build, but anyone that have the rights to do, go ahead.

francispoulin commented:
Hello Alexis (@amontoison), now that this has merged, any chance we could chat about how to use this library?  My email is fpoulin@uwaterloo.ca.  Thanks in advance!
